nixos/paperless: fix start with latest systemd

leona-ya opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

After the last systemd update, paperless didn't start at all (see linked issue for details). ~~I'm not sure whether this is the best approach, but imho it should be fixed, and then we can find the best solution.~~

fixes  #258823

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Also tested manually with postgresql.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

leona-ya committed:
nixos/paperless: fix start with latest systemd

andir commented on nixos/modules/services/misc/paperless.nix:
```diff
@@ -71,7 +71,8 @@ let
     #ProtectSystem = "strict";
-    ProtectControlGroups = true;
+    # Broke with systemd 254
+    #ProtectControlGroups = true;
```
This is likely the wrong fix to apply here. The acutal issue, according to my conversation with upstream systemd (https://github.com/systemd/systemd/issues/29449) is that `/run` doesn't exist anymore (and that our `TemporaryFilesystem=/:ro` cargo culting is just wrong given that we have `ProtectSystem=` available).

leona-ya replied:
Do you have an opinion whether to use `ProtectSystem=strict` with `ReadWritePaths=` or just `ProtectSystem = true`?

andir replied:
I think we can (as much as we can) stick with `ProtectSystem=strict` and provide the `ReadWritePaths` (or the bind equivalent) as much as possible. Not defining every path and letting systemd take care of the "system protection" (in the case of `ProtectSystem`) is likely the better way going forward. It is more upgradeable in terms of security settings than what we've done with defining e.g. `/` as `ro` etc..

andir reviewed (commented):
The test passes for me. The changes look good but I'd like to ensure that we aren't breaking the optional redis support that previously had some unit configurations. Can you add that as another configuration within the same nixos test or add another?

Flakebi commented:
Just to mention it, IMO `ProtectSystem=true` provides less isolation than `TemporaryFileSystem=/`, because it bind-mounts all kinds of paths and thus makes them visible to the service. For example all of `/etc` and `/var/lib` are now visible to paperless, before it could only see the things it actually needed.

leona-ya committed:
nixosTests.paperless: add additional test using postgresql as db

With that test we also test, that paperless can connect to the outside
postgresql socket.

Co-Authored-By: Erik Arvstedt <erik.arvstedt@gmail.com>

leona-ya commented:
> The test passes for me. The changes look good but I'd like to ensure that we aren't breaking the optional redis support that previously had some unit configurations. Can you add that as another configuration within the same nixos test or add another?

Actually redis support is already tested: Redis is not optional, but necessary for the task queue. I added an additional test, tetsing that the optional postgres connection works.

andir commented:
> Just to mention it, IMO `ProtectSystem=true` provides less isolation than `TemporaryFileSystem=/`, because it bind-mounts all kinds of paths and thus makes them visible to the service. For example all of `/etc` and `/var/lib` are now visible to paperless, before it could only see the things it actually needed.

I was also able to see the whole of the Nix store. Not just what it needed. The bind mounts we had might be less surface than what this option provides but IMO we should either invest in proper composable isolation for services or stick with the systemd facilities. Cargo-culting a list of options from one service to the next will continuously run into these kinds of issues. Permissions should be opt-in and not opt-out :(

erikarvstedt commented:
Thanks for looking into this! [Some fixups](https://github.com/erikarvstedt/nixpkgs/commits/nixos-paperless-systemd-254-ea).

Flakebi reviewed (approved):
LGTM as well, thanks for fixing the service!

erikarvstedt commented on nixos/tests/paperless.nix:
```diff
@@ -2,65 +2,88 @@ import ./make-test-python.nix ({ lib, ... }: {
-      enable = true;
-      passwordFile = builtins.toFile "password" "admin";
+  nodes = let self = {
+    simple = { pkgs, ... }: {
+      environment.systemPackages = with pkgs; [ imagemagick jq ];
+      services.paperless = {
+        enable = true;
+        passwordFile = builtins.toFile "password" "admin";
+      };
     };
-  };
+    postgres = { config, pkgs, ... }: {
+      imports = [ self.simple ];
+      services.postgresql = {
+        enable = true;
+        ensureDatabases = [ "paperless" ];
+        ensureUsers = [
+          { name = config.services.paperless.user;
+            ensurePermissions = { "DATABASE \"paperless\"" = "ALL PRIVILEGES"; };
+          }
+        ];
+      };
+      services.paperless.extraConfig = {
+        PAPERLESS_DBHOST = "/run/postgresql";
+      };
+    };
```
I introduced this change. `rec` is a nix antipattern that should be avoided in larger attrsets ([1](https://nix.dev/recipes/best-practices#recursive-attribute-set-rec), [2](https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293)). But I have no strong preference here.

Atemu replied:
Note that nix.dev is an opinionated guide; there is no broader consensus on `rec` that I'm aware of. What I believe everyone can agree with though is that `rec` can be confusing when used in expressions that can be overridden such as mkDerivation but, while an alternative exists, it has performance drawbacks, so even there isn't a clear-cut reason to avoid rec. This isn't the case here however.

I personally think a rec is easier to read in many cases but I'm not totally against patterns like this; it just stood out as a bit weird to me.

Atemu commented:
@ofBorg build paperless-ngx.tests
